File: Data Structures & Algorithms/validate-parentheses/submission-5.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) < 2:
            return False

        st = []
        
        st.append(s[0])
        for i in range(1, len(s)):
            if s[i] == '(' or s[i] == '[' or s[i] == '{':
                st.append(s[i])
            elif s[i] == ')':
                if len(st) == 0:
                    return False
                if st[-1] != '(':
                    return False
                st.pop()
            elif s[i] == '}':
                if len(st) == 0:
                    return False
                if st[-1] != '{':
                    return False
                st.pop()
            elif s[i] == ']':
                if len(st) == 0:
                    return False
                if st[-1] != '[':
                    return False
                st.pop()

        return len(st) == 0

File: Data Structures & Algorithms/clone-graph/submission-5.java


/*
Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/


class Solution {

    Node getNewNode(int val){
        return new Node(val, new ArrayList<Node>());
    }
    public Node cloneGraph(Node node) {
        if (node == null) {
            return node;
        }

        // also acts as seen node
        Map<Node, Node> visited = new HashMap<>();

        Deque<Node> st = new ArrayDeque<>();
        st.push(node);
        visited.put(node, getNewNode(node.val));

        while (!st.isEmpty()) {
            Node curr = st.pop();
            Node currClone = visited.get(curr);

            for (Node nei : curr.neighbors) {
                if (!visited.containsKey(nei)) {
                    visited.put(nei, getNewNode(nei.val));
                    st.push(nei);
                }
                // add neighbors if already cloned before
                // don't re-clone
                currClone.neighbors.add(visited.get(nei));
            }
        }
        return visited.get(node);
    }
}